### PR TITLE
feat: implement resource and data sources to create and fetch direct link route reports

### PR DESCRIPTION
**Pull request by ajay-malhotra1**
Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccIBMDLRouteReportDataSource_basic'
=== RUN   TestAccIBMDLRouteReportDataSource_basic
--- PASS: TestAccIBMDLRouteReportDataSource_basic (309.57s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink	310.318s
...
```

```
$ make testacc TESTARGS='-run=TestAccIBMDLRouteReportResource_basic'
=== RUN   TestAccIBMDLRouteReportResource_basic
--- PASS: TestAccIBMDLRouteReportResource_basic (281.03s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink	281.729s
...
```

```
$ make testacc TESTARGS='-run=TestAccIBMDLRouteReportsDataSource_basic'
=== RUN   TestAccIBMDLRouteReportsDataSource_basic
--- PASS: TestAccIBMDLRouteReportsDataSource_basic (356.05s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink	356.788s
...
```

### CODE REVIEW

File: ibm/service/directlink/resource_ibm_dl_route_report.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+func resourceIBMdlGatewayRouteReportCreate(d *schema.ResourceData, meta interface{}) error {
+	directLink, err := directlinkClient(meta)
+	if err != nil {
+		return err
+	}
+
+	gatewayId := d.Get(dlGatewayId).(string)
+	createGatewayRouteReportOptionsModel := &directlinkv1.CreateGatewayRouteReportOptions{GatewayID: &gatewayId}
+	routeReport, response, err := directLink.CreateGatewayRouteReport(createGatewayRouteReportOptionsModel)
+	if err != nil {
```
Comment (kavya498) [inline ibm/service/directlink/resource_ibm_dl_route_report.go:181]: check for routeReport == nil along with err !=nil

File: ibm/service/directlink/resource_ibm_dl_route_report.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+	gatewayId := parts[0]
+	routeReportId := parts[1]
+
+	log.Println("[Info] Fetching DL Route Report GW ID:", gatewayId, " and Report ID: ", routeReportId)
+
+	getGatewayRouteReportOptionsModel := &directlinkv1.GetGatewayRouteReportOptions{GatewayID: &gatewayId, ID: &routeReportId}
+	report, response, err := directLink.GetGatewayRouteReport(getGatewayRouteReportOptionsModel)
+	if err != nil {
+		log.Println("[WARN] Error fetching DL Route Reports", response, err)
```
Comment (kavya498) [inline ibm/service/directlink/resource_ibm_dl_route_report.go:252]: check for response status == 404.. if 404 setId to "". This helps in lifecycle management.
Refer other resources for reference

File: metadata/provider_metadata.json [hunk 3]
```diff
@@ -17554,6 +17554,257 @@
                 }
             }
         ],
+        "ibm_dl_route_report": [
```
Comment (kavya498) [inline metadata/provider_metadata.json:17557]: This file is auto-generated during release.. Kindly remove changes from this file

File: website/docs/r/dl_route_report.html.markdown [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  type =  "connect"
+  port =  data.ibm_dl_ports.test_ds_dl_ports.ports[0].port_id
+}
+
+resource ibm_dl_route_report dl_route_report {
+    gateway = ibm_dl_gateway.test_dl_gateway.id
+}
+```
+
+## Argument reference
```
Comment (kavya498) [inline website/docs/r/dl_route_report.html.markdown:65]: Please make sure argument reference matches with resource schema.
Any Required / Optional / Optional and Computed arguments in the schema should be mentioned in Argument reference section

Any argument that is marked as only computed in schema should be mentioned in Attribute reference..

All

File: website/docs/r/dl_route_report.html.markdown [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    Nested scheme for `virtual_connection_routes`
+    - `routes` - (List) List of connection routes.
+        Nested scheme for `routes`:
+        - `prefix` - (String) The prefix used in the route.
+    - `virtual_connection_id` - (String) Virtual Connection ID
+    - `virtual_connection_name` - (String) Virtual Connection name
+    - `virtual_connection_type` - (String) Virtual Connection type
+
+## Import
+The `ibm_dl_gateway` resource can be imported by using gateway ID.
```
Comment (kavya498) [inline website/docs/r/dl_route_report.html.markdown:99]: Import is possible with the id that is set for d.SetId in the create method of the resource..

Id is set to a combination in the resource
```
d.SetId(fmt.Sprintf("%s/%s", gatewayId, *routeReport.ID))
```

Resource can be imported using <gateway_ID>/<report_ID>. Please update the docs accordingly.

File: website/docs/r/dl_route_report.html.markdown [hunk 6]
```diff
@@ -0,0 +1,112 @@
+---
+subcategory: "Direct Link Gateway"
+layout: "ibm"
+page_title: "IBM : dl_route_report"
+description: |-
+  Manages IBM Direct Link Gateway.
```
Comment (kavya498) [inline website/docs/r/dl_route_report.html.markdown:6]: Kindly change to appropriate description of the resource

**Review by kavya498 — CHANGES_REQUESTED** [review]
Kindly address review comments, rebase and resolve conflicts
Thanks

**Comment by ajay-malhotra1** [comment]
@kavya498 The review comments have been addressed. Can you please review it again

File: ibm/service/directlink/data_source_ibm_dl_route_reports_test.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		speed_mbps = 1000 
+		type =  "connect" 
+		port = data.ibm_dl_ports.ds_dlports.ports[0].port_id
+	} 
+
+	resource ibm_dl_route_report dl_route_report {
+		gateway = ibm_dl_gateway.test_dl_gateway.id
+	 }
+
+	 data "ibm_dl_route_reports" "test_dl_reports" {
```
[the remark this answers is not in the document]
Comment (ajay-malhotra1) [reply ibm/service/directlink/data_source_ibm_dl_route_reports_test.go:53]: This datasource is referencing the resource ibm_dl_route_report now and will be called after that route report resource creation.

File: ibm/service/directlink/data_source_ibm_dl_route_report.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+	}
+
+	gatewayId := d.Get(dlGatewayId).(string)
+	routeReportId := d.Get(dlRouteReport).(string)
+
+	log.Println("[Info]  fetching DL Route Reports GW ID:", gatewayId, " and report ID: ", routeReportId)
+
+	getGatewayRouteReportOptionsModel := &directlinkv1.GetGatewayRouteReportOptions{GatewayID: &gatewayId, ID: &routeReportId}
+	report, response, err := directLink.GetGatewayRouteReport(getGatewayRouteReportOptionsModel)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (ajay-malhotra1) [reply ibm/service/directlink/data_source_ibm_dl_route_report.go:170]: Added the `nil` check for `report` separately below to return a different error

File: ibm/service/directlink/resource_ibm_dl_route_report.go (hunk 1 above)
Comment (ajay-malhotra1) [reply ibm/service/directlink/resource_ibm_dl_route_report.go:173]: Added the `nil` check for `routeReport` separately below to return a different error

**Pull request merged**